[ao][sparsity] comsability for sparsity and QAT convert

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74848

Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations
to test_nn.py and also added comments to the test code for
composability.

Test Plan: python test/test_ao_sparsity.py TestComposability
python test/test_nn.py TestNN

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

HDCharles committed:
composability sparsity+QAT

Summary: WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "composability sparsity+QAT"


Summary: WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/nn/qat/modules/linear.py:
```diff
@@ -38,17 +40,25 @@ def from_float(cls, mod):
             or directly from user
         """
-        assert type(mod) == cls._FLOAT_MODULE, ' qat.' + cls.__name__ + '.from_float only works for ' + \
+        from torch.ao.quantization.utils import nonparam_type  # import error if this is moved to the top
```
If anyone knows a good way to avoid the import error or otherwise make this cleaner, let me know.

vkuzo replied:
maybe we can just move this util out of `torch.ao` and into `torch.nn.utils.parametrize`?  The function itself is not related to quantization, it's just about parametrizations

HDCharles replied:
are we allowed to do that? I am generally wary of messing with other teams' code

vkuzo replied:
it sounds reasonable to me in this case, but perhaps we can also ask the folks working on parametrization code?  I see that @albanD  and @soulitzer have committed code to that directory recently, maybe they have some thoughts?

albanD replied:
I'm not sure what that function is doing. But if it is an API that could be generally useful we're always ok to add it.

To avoid the import issue can you just use `torch.ao.quantization.utils.nonparam_type` here directly?

HDCharles replied:
I can it just makes things really hard to read

HDCharles replied:
i'll add this and the transfer one into torch.nn.utils.parametrize

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] composability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

vkuzo commented on torch/ao/quantization/utils.py:
```diff
@@ -384,3 +385,11 @@ def is_leaf_or_only_parametrized(module):
     else:
         return False
+
+def transfer_parametrizations_and_params(from_mod, to_mod):
```
optional: this looks like a generic util unrelated to quantization, we could put it in parametrization code so others could use it as well

albanD commented on torch/ao/quantization/utils.py:
```diff
@@ -372,7 +373,7 @@ def type_before_parametrizations(module):
         return type(module)
 
-def is_leaf_or_only_parametrized(module):
+def has_no_children_ignoring_parametrizations(module):
```
Just a side note. But if you move functions to the parametrize module, then we do enforce the FC/BC rules. In particular, you are not allowed to just rename a public function. It should be deprecated first and then deleted. Details here: https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy

HDCharles replied:
Yeah, my git-fu is not as strong as it could be, I meant for that to be changed to be in the base commit which is now rectified. The function didn't exist before this stack so hopefully, after moving the creation of the function into the base commit this satisfies the BC requirements.

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -655,3 +655,19 @@ def type_before_parametrizations(module: Module) -> type:
+        assert isinstance(from_module.parametrizations, ModuleDict)
+        for parameter_name in from_module.parametrizations:
+            for param_func in from_module.parametrizations[parameter_name]:
+                setattr(to_module, parameter_name, from_module.parametrizations[parameter_name].original)
```
I agree. This should be done using the current API. I think this, as implemented, should break, at least, caching.

HDCharles replied:
@Lezcano @albanD can you help me understand the issue, this function isn't removing any parametrizations, its just duplicating them onto a new module

Lezcano replied:
The caching system registers the module id and the tensor id into a cache and remembers the result.

https://github.com/pytorch/pytorch/blob/a8e45b596910f90013637ccc735d6ca21a93d852/torch/nn/utils/parametrize.py#L321

It's not clear to me whether this would work after calling `transfer_parametrizations_and_params`.

Furthermore, I don't think this is really correct. For once, you are registering the parametrizations from the outside to the inside, which is the inverse order they were added in the `from_module`.

Could you add some tests for this function that tests it when registering more than one parametrisation and test the caching as well?

I don't understand the `setattr` either, as the parameters within the parametrisation would also be reachable within the `to_module` by going through `to_module.parametrizations[parameter_name].original`. The tests should also test that this is not the case.

HDCharles replied:
OK, I have added the tests and verified that it doesn't break caching and that the ordering seems to be correct. I am not understanding the last paragraph of your comment though @Lezcano.

> don't understand the setattr either, as the parameters within the parametrisation would also be reachable within the to_module by going through to_module.parametrizations[parameter_name].original. The tests should also test that this is not the case.

I'm unsure why that is odd and/or needs to be tested. If you use any other method to transfer a weight or something from one module to another, isn't that how you would expect it to behave?

i.e. if you do something like

```
lin = torch.nn.Linear(in, out)
new_lin = torch.nn.qat.Linear(in, out, qconfig)
new_lin.weight = lin.weight
```

My goal was to do that, but transfer the value and the parametrizations as well.

I updated the transfer function a bit to match this intent.

Lezcano replied:
Great to know that everything's working as expected! I'll go over the tests later today. Sorry for the hassle :)

My point here is that, what you're doing at the moment is to set `other.weight = original`, and then you're registering parametrisations on top. Registering these parametrizations, if they have `right_inverse`s, would change that weight. What you should do is to set the attr to `getattr(from_module, parameter_name)` I believe. Now, if `not is_parametrized(module, parameter_name)`, you can also do it the other way around: You first register the parametrization, and then do `to_module.parametrizations[parameter_name].original = from_module.parametrizations[parameter_name].original`, or something of sorts. This latter approach would be preferred whenever possible for numerical stability reasons.

HDCharles replied:
OK I will investigate, is there a better test example I can use to verify that I'm doing it correctly once I fix this or should I just look for any transform with a right inverse?

Lezcano replied:
Already using one parametrisation with a `right_inverse` should be enough to show the problem. If that test passes I think that this should be ready to go.

HDCharles replied:
oh yeah, thanks for identifying this, that bug existed and on top of that a worse bug existed where the original module would get changed when transferring the parametrization

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

also added comments to the test code

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module.

also added comments to the test code

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to identify all
parametrizations on the original module and then move them (and their
associated parameters) to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations

also added comments to the test code

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations

also added comments to the test code

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

Lezcano commented on test/test_nn.py:
```diff
@@ -3171,6 +3171,37 @@ def forward(self, X):
+
+            A = to_model.weight
+            B = to_model.weight
+            self.assertEqual(id(X), id(Y))
```
Ooops. 
```suggestion
            self.assertEqual(id(A), id(B))
```
You should probably also add a `self.assertEqual(id(A), id(X))`, right?

HDCharles replied:
I don't think so, wouldn't that only happen if the parametrizationList is the same object for both, rather than a copy? That wasn't my intent so it should cache these seperately, right? i.e. if (after the transfer) you remove the parametrization from the from_module it won't remove it from the to_module. (I added a test for this last bit just to double check)

Lezcano replied:
So, the cache works using the id of the module + the original tensor. I assume that, since you're doing a deepcopy, these two are indeed cached as different objects. Now, if this is the intended behaviour, I would suggest to add the opposite test, that is `self.assertNotEqual(id(A), id(X))`

HDCharles replied:
oh yeah thats a good idea

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations

also added comments to the test code

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations

also added comments to the test code

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations
to test_nn.py and also added comments to the test code for
composability.

Test Plan: python test/test_ao_sparsity.py TestComposability
python test/test_nn.py TestNN

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

lezcano reviewed (commented):
Just one more thing I forgot! Parametrisations may be many-to-one if the right-inverse returns more than one tensor. You can find examples of these in `test_multiple_inputs_parametrization`. Could you add a test for this case?

HDCharles commented on torch/nn/utils/parametrize.py:
```diff
@@ -648,10 +648,44 @@ def type_before_parametrizations(module: Module) -> type:
+
+            # need to initialize the param in to_module if it doesn't exist already
+            if not hasattr(to_module, parameter_name):
+                setattr(to_module, parameter_name, deepcopy(from_module.parametrizations[parameter_name].original))
```
yeah, its tested in line 3282 in test_nn.py within test_transfer_parametrizations_and_params, I can split it into another test if you like, it looked like some of the other parametriation tests were grouped in some similar cases so I wasn't sure.

I added your suggested change, and added a test for the many-to-one case which works after some changes.

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations
to test_nn.py and also added comments to the test code for
composability.

Test Plan: python test/test_ao_sparsity.py TestComposability
python test/test_nn.py TestNN

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations
to test_nn.py and also added comments to the test code for
composability.

Test Plan: python test/test_ao_sparsity.py TestComposability
python test/test_nn.py TestNN

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).

lezcano reviewed (approved):
LGTM CI abiding. Thank you for the many improvements and the testing on the parametrisations end!

HDCharles committed:
Update on "[ao][sparsity] comsability for sparsity and QAT convert"


Summary: The primary issue for enabling sparsity to work with QAT
convert (unlike normal quantization convert) is that when the
parametrized module undergoes the QAT convert, the parametrizations need
to be maintained. If the parametrizations don't
get transfered during the convert, the sparsifier would lose its
connection to the model. In practice this was handled using the
transfer_parametrizations_and_params function to move the weight and
bias and any associated paramerizations to the new module. This PR also adds
tests for transfer_parametrizations_and_params and type_before_parametrizations
to test_nn.py and also added comments to the test code for
composability.

Test Plan: python test/test_ao_sparsity.py TestComposability
python test/test_nn.py TestNN

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240272](https://our.internmc.facebook.com/intern/diff/D35240272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240272).
